doc/src/explanation/glossary.md

```diff
@@ -0,0 +1,712 @@
+    very specific mathematical manipulations to an expression.
+
+Solve
+Solvers
+
+    To *solve* an {term}`equation` or system of equations means to find a set
+    of {term}`expressions <expression>` that make the equation(s) true when
+    the given {term}`symbol(s) <symbol>` are {term}`substituted
+    <substitution>` with them. For example, the solution to the equation $x^2
+    = 1$ with respect to $x$ would be the set $\{-1, 1\}$. Different types of
+    equations can be solved by SymPy using different
+    [*solvers*](solving-guide) functions. For instance, algebraic equations
+    can be solved with {func}`~.solve`, differential equations can be solved
+    with {func}`~.dsolve`, and so on.
+
+    SymPy generally uses the word "solve" and "solvers" to mean equation
+    solving in this sense. It is not used in the sense of "solving a problem".
+    For instance, one would generally prefer to say "compute an integral" or
+    "evaluate an integral" rather than "solve an integral" to refer to
+    symbolic integration using the function {func}`~.integrate`.
+
+Structural Equality
+
+    Two SymPy objects are *structurally equal* if they are equal as
+    {term}`expressions <expression>`, that is, they have the same as
+    {term}`expression trees <expression tree>`. Two structurally equal
+    expressions are considered to be identical by SymPy, since all SymPy
+    expressions are {term}`immutable`. Structural equality can be checked with
+    the `==` operator, which always returns `True` or `False`. Symbolic
+    {term}`equality <equation>` can be represented using {class}`Eq
+    <sympy.core.relational.Equality>`.
+
+    Typically, two expressions are structurally equal if they are the same
+    class and (recursively) have the same {term}`args`. Two expressions may be
+    mathematically identical but not structurally equal. For example, `(x +
+    1)**2` and `x**2 + 2*x + 1` are mathematically equal, but they are not
+    structurally equal, because the first is a {class}`~.Pow` whose
+    {term}`args` consist of an {class}`~.Add` and an {class}`~.Integer`, and
+    the second is an {class}`~.Add` whose {term}`args` consist of a
+    {class}`~.Pow`, a {class}`~.Mul`, and an {class}`~.Integer`.
+
+    Two apparently different expressions may be structurally equal if they are
+    {term}`canonicalized <canonicalize>` to the same thing by {term}`automatic
+    simplification`. For example, `x + y` and `y + x` are structurally equal
+    because the {class}`~.Add` constructor automatically sorts its arguments,
+    making them both the same.
+
+Subexpression
+
+    A *subexpression* is an {term}`expression` that is contained within a
+    larger expression. A subexpression appears somewhere in the
+    {term}`expression tree`. In some contexts, a subtree of the expression
+    tree is also considered a subexpression. For instance, `x + y` may
+    sometimes be considered a subexpression of `x + y + z`, even though the
```

Comment:
I've never seen the word "subtree" used in this way.

Reply:
You mean the word "subtree" to refer to a subset of a (mathematical) tree? I would assume that's common, but perhaps I can find a reference (or more commonly used term).

Or do you mean you haven't seen `x + y` referred to as a subexpression of `x + y + z`? I see that a lot, especially in functions like `subs` that operate at a level higher than pure expression trees.

Reply:
At the bottom of this wikipedia page there is a footnote:

This is different from the formal definition of subtree used in graph theory, which is a subgraph that forms a tree – it need not include all descendants. For example, the root node by itself is a subtree in the graph theory sense, but not in the data structure sense (unless there are no descendants).
https://en.wikipedia.org/wiki/Tree_(data_structure)

I'm not sure that the word subtree as used here corresponds to either of those definitions though:

- The data structure subtree would not allow `x + y` to be a subtree of `x + y + z`.
- The graph theory subtree would allow `x + y*z` to be a subtree of `x + y*z*t`.

Reply:
Really this only applies to the associative operators Add and Mul, where you can take subsets of them and consider that to be a "subexpression". You wouldn't consider `f(x, y)` to be a subexpression of `f(x, y, z)` for any other `f`. So maybe instead of trying to describe this in terms of graph theory, it can just mention that sometimes commutative or associative laws are taken into account for deciding what is a subexpression, which can expand past the literal "subtree" definition.

We really should have more formal ways of thinking about this sort of thing in SymPy, along the lines of MatchPy's commutative/associative pattern matching (and I think e.g., Mathematica has similar notions). But presently it's all treated in an ad hoc manner.

Reply:
The examples I showed use only associative, commutative operations. I think what you are describing is like the data structure subtree except that only a subset of children needs to be taken at the second level of the tree (the level below the outer Add).